File: cpp/src/Ice/WSConnector.cpp
// Copyright (c) ZeroC, Inc.

#include "WSConnector.h"
#include "HttpParser.h"
#include "TargetCompare.h"
#include "WSEndpoint.h"
#include "WSTransceiver.h"

using namespace std;
using namespace Ice;
using namespace IceInternal;

TransceiverPtr
IceInternal::WSConnector::connect()
{
    return make_shared<WSTransceiver>(_instance, _delegate->connect(), _host, _resource);
}

int16_t
IceInternal::WSConnector::type() const
{
    return _delegate->type();
}

string
IceInternal::WSConnector::toString() const
{
    return _delegate->toString();
}

bool
IceInternal::WSConnector::operator==(const Connector& r) const
{
    const auto* p = dynamic_cast<const WSConnector*>(&r);
    if (!p)
    {
        return false;
    }

    if (this == p)
    {
        return true;
    }

    if (_resource != p->_resource)
    {
        return false;
    }

    return Ice::targetEqualTo(_delegate, p->_delegate);
}

bool
IceInternal::WSConnector::operator<(const Connector& r) const
{
    const auto* p = dynamic_cast<const WSConnector*>(&r);
    if (!p)
    {
        return type() < r.type();
    }

    if (this == p)
    {
        return false;
    }

    if (_resource < p->_resource)
    {
        return true;
    }
    else if (p->_resource < _resource)
    {
        return false;
    }

    return Ice::targetLess(_delegate, p->_delegate);
}

IceInternal::WSConnector::WSConnector(ProtocolInstancePtr instance, ConnectorPtr del, string host, string resource)
    : _instance(std::move(instance)),
      _delegate(std::move(del)),
      _host(std::move(host)),
      _resource(std::move(resource))
{
}

IceInternal::WSConnector::~WSConnector() = default;

File: java-compat/src/Ice/src/main/java/IceInternal/WSConnector.java
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

package IceInternal;

final class WSConnector implements Connector
{
    @Override
    public Transceiver connect()
    {
        return new WSTransceiver(_instance, _delegate.connect(), _host, _resource);
    }

    @Override
    public short type()
    {
        return _delegate.type();
    }

    @Override
    public String toString()
    {
        return _delegate.toString();
    }

    @Override
    public int hashCode()
    {
        return _delegate.hashCode();
    }

    WSConnector(ProtocolInstance instance, Connector del, String host, String resource)
    {
        _instance = instance;
        _delegate = del;
        _host = host;
        _resource = resource;
    }

    @Override
    public boolean equals(java.lang.Object obj)
    {
        if(!(obj instanceof WSConnector))
        {
            return false;
        }

        if(this == obj)
        {
            return true;
        }

        WSConnector p = (WSConnector)obj;
        if(!_delegate.equals(p._delegate))
        {
            return false;
        }

        if(!_resource.equals(p._resource))
        {
            return false;
        }

        return true;
    }

    private ProtocolInstance _instance;
    private Connector _delegate;
    private String _host;
    private String _resource;
}
